Add (broken interleaved) paired-end support to sample-reads-randomly.py script. Closes #899.

standage opened the pull request:
- [x] Is it mergeable?
- [x] Did it pass the tests?
- [x] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [x] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

standage commented:
Note: I've used the broken paired reader for paired-end support. Pairs are treated as a single unit and will either both be output, or neither. In a file containing only paired-end reads, requesting 10 reads will give you 20 records (10 record pairs). In a file containing only single-end reads, requesting 10 reads will give you 10 records. In a broken interleaved file, what you get will depend on the sampling.

standage commented:
Ready for code review.

standage commented:
Closes #899.

standage commented:
Poke @mr-c

mr-c commented:
Punting to @ctb

ctb commented on scripts/sample-reads-randomly.py:
```diff
@@ -158,7 +162,9 @@ def main():
 
         for record in reads[0]:
-            write_record(record, output_file)
+            write_record(record[2], output_file)
```
not critical, but for clarity might suggest

    for (record1, record2) in reads[0]: ...

ctb commented on scripts/sample-reads-randomly.py:
```diff
@@ -127,7 +130,8 @@ def main():
         print >>sys.stderr, 'opening', filename, 'for reading'
-        for record in screed.open(filename, parse_description=False):
+        screed_iter = screed.open(filename, parse_description=False)
+        for record in broken_paired_reader(screed_iter):
```
I personally dislike treating 'record' as opaque here -- and throughout the code base, 'record' is used as a single screed record. so would suggest breaking this out into _, is_paired, record1, record2.

ctb replied:
(or trying out a new convention that doesn't use 'record')

ctb commented on scripts/sample-reads-randomly.py:
```diff
@@ -63,6 +63,9 @@ def get_parser():
     parser.add_argument('-S', '--samples', type=int, dest='num_samples',
                         default=1)
     parser.add_argument('-R', '--random-seed', type=int, dest='random_seed')
+    parser.add_argument('-p', '--paired', default=False, dest='paired',
```
This option isn't used at all, is it?

standage commented:
Correct, I have removed the option.

standage commented:
@ctb Agree with your comments regarding "records." I like to think of a record as a single entry in a fastq file, and this code abused that terminology. I was leaning toward unpacking the return values of **broken_paired_reader** as soon as I committed the code, but alas, travel. :(

standage commented:
Ready for code review @ctb.

ctb commented on ChangeLog:
```diff
@@ -9,6 +9,13 @@
 
 2015-04-13  Daniel Standage  <daniel.standage@gmail.com>
 
+   * scripts/sample-reads-randomly.py: use broken paired reader to provide paired-
```
Please move this to the top now ;)

ctb replied:
(and update the date)

ctb replied:
Also, make sure lines are wrapped before 80 characters, please; go ahead and fix Elmar Bucher's 4/13 entry, too, if you can.

ctb commented:
LGTM outside of the changelog fixes.  One last question - what about providing a command line option to force single-ended behavior? (see force_single argument to broken_paired_reader). I don't see a reason to force paired-ends (there are other ways to do that) but the user may well want to excplitly ignore pairs when sampling, which there is no way to do currently in the code base or in this code.

standage commented:
I agree that forcing paired wouldn't make much sense, but ignoring pair information is a valid use case. Since it's already implemented as an option to broken_paired_reader, it should be easy to add an option to activate this mode. I'll fix the change log, add the option, and then (when I have time) add a test for the new option.

ctb commented:
thanks much appreciated :)

standage commented:
Huh. For some reason the script is reporting N+1 reads / read pairs instead of N. Checking whether I introduced a fencepost error...

standage commented:
Jenkins retest this please.

standage commented:
- [x] New `--force_single` option added, with corresponding test
- [x] Fixed changelog
- [x] Fixed pep8 errors introduced by recently merged commits

Ready for final review, @ctb.

ctb commented:
I think we use - in options, so --force-single. ;)

--
Titus Brown, ctbrown@ucdavis.edu

> On Apr 21, 2015, at 9:46 PM, Daniel Standage <notifications@github.com> wrote:
> 
>  New --force_single option added, with corresponding test
>  Fixed changelog
>  Fixed pep8 errors introduced by recently merged commits
> Ready for final review, @ctb.
> 
> —
> Reply to this email directly or view it on GitHub.
>

standage commented:
There is some inconsistency in this script: `--num_reads`, `--max_reads`, and `--random-seed`. I can fix this new option, but fixing the existing options would require a major version bump, would it not?

ctb commented:
Ok leave as is - will review tomorrow morning!

> On Apr 21, 2015, at 10:01 PM, Daniel Standage <notifications@github.com> wrote:
> 
> There is some inconsistency in this script: --num_reads, --max_reads, and --random-seed. I can fix this new option, but fixing the existing options would require a major version bump, would it not?
> 
> —
> Reply to this email directly or view it on GitHub.
>

ctb commented:
LGTM; thanks, @standage !
